File: flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java
/*
 * Copyright 2021 Flyte Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.flyte.flytekit;

import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toUnmodifiableList;
import static java.util.stream.Collectors.toUnmodifiableMap;
import static org.flyte.flytekit.SdkLiteralTypes.collections;
import static org.flyte.flytekit.SdkLiteralTypes.maps;

import com.google.auto.value.AutoValue;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import org.flyte.api.v1.BindingData;
import org.flyte.api.v1.LiteralType;
import org.flyte.api.v1.OutputReference;

/** Specifies either a simple value or a reference to another output. */
public abstract class SdkBindingData<T> {

  abstract BindingData idl();

  public abstract SdkLiteralType<T> type();

  /**
   * Returns the simple value contained by this data.
   *
   * @return the value that this simple data holds
   * @throws IllegalArgumentException when this data is an output reference
   */
  public abstract T get();

  /**
   * Returns a version of this {@code SdkBindingData} with a new type.
   *
   * @param newType the {@link SdkLiteralType} type to be casted to
   * @param castFunction function to apply to the value to be converted to the new type
   * @return the type casted version of this instance
   * @param <NewT> the java or scala type for the corresponding to {@code newType}
   * @throws UnsupportedOperationException if a cast cannot be performed over this instance.
   */
  public abstract <NewT> SdkBindingData<NewT> as(
      SdkLiteralType<NewT> newType, Function<T, NewT> castFunction);

  /**
   * Creates a {@code SdkBindingData} for a literal value.
   *
   * @param type the {@link SdkLiteralType} type
   * @param value contains the simple value of this class
   * @return A newly created SdkBindingData
   * @param <T> the java or scala type for the corresponding LiteralType, for example {@code
   *     Duration} for {@code LiteralType.ofSimpleType(SimpleType.DURATION)}
   */
  public static <T> SdkBindingData<T> literal(SdkLiteralType<T> type, T value) {
    return Literal.create(type, value);
  }

  /**
   * Creates a {@code SdkBindingData} for a reference to (promise for) another output.
   *
   * @param type the {@link SdkLiteralType} type
   * @param nodeId which nodeId to reference
   * @param var variable name to reference on the node id
   * @return A newly created SdkBindingData
   * @param <T> the java or scala type for the corresponding LiteralType, for example {@code
   *     Duration} for {@code LiteralType.ofSimpleType(SimpleType.DURATION)}
   */
  public static <T> SdkBindingData<T> promise(SdkLiteralType<T> type, String nodeId, String var) {
    return Promise.create(type, nodeId, var);
  }

  /**
   * Creates a {@code SdkBindingData} for a collections of {@link SdkBindingData}.
   *
   * @param elementType the {@link SdkLiteralType} of the elements of the collection.
   * @param collection collections of {@link SdkBindingData}s
   * @return A newly created SdkBindingData
   * @param <T> the java or scala type for the corresponding LiteralType, for example {@code
   *     Duration} for {@code LiteralType.ofSimpleType(SimpleType.DURATION)}
   */
  public static <T> SdkBindingData<List<T>> bindingCollection(
      SdkLiteralType<T> elementType, List<SdkBindingData<T>> collection) {
    return BindingCollection.create(elementType, collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a map of {@link SdkBindingData}.
   *
   * @param valuesType the {@link SdkLiteralType} of the elements of the collection.
   * @param map map of {@link SdkBindingData}s
   * @return A newly created SdkBindingData
   * @param <T> the java or scala type for the corresponding LiteralType, for example {@code
   *     Duration} for {@code LiteralType.ofSimpleType(SimpleType.DURATION)}
   */
  public static <T> SdkBindingData<Map<String, T>> bindingMap(
      SdkLiteralType<T> valuesType, Map<String, SdkBindingData<T>> map) {
    return BindingMap.create(valuesType, map);
  }

  @AutoValue
  abstract static class Literal<T> extends SdkBindingData<T> {
    abstract T value();

    private static <T> Literal<T> create(SdkLiteralType<T> type, T value) {
      return new AutoValue_SdkBindingData_Literal<>(type, value);
    }

    @Override
    BindingData idl() {
      return type().toBindingData(value());
    }

    @Override
    public T get() {
      return value();
    }

    @Override
    public <NewtT> SdkBindingData<NewtT> as(
        SdkLiteralType<NewtT> newType, Function<T, NewtT> castFunction) {
      return create(newType, castFunction.apply(value()));
    }

    @Override
    public final String toString() {
      return String.format("SdkBindingData{type=%s, value=%s}", type(), value());
    }
  }

  @AutoValue
  abstract static class Promise<T> extends SdkBindingData<T> {
    abstract String nodeId();

    abstract String var();

    private static <T> Promise<T> create(SdkLiteralType<T> type, String nodeId, String var) {
      return new AutoValue_SdkBindingData_Promise<>(type, nodeId, var);
    }

    @Override
    BindingData idl() {
      return BindingData.ofOutputReference(
          OutputReference.builder().nodeId(nodeId()).var(var()).build());
    }

    @Override
    public T get() {
      throw new IllegalArgumentException(
          String.format(
              "Value only available at workflow execution time: promise of %s[%s]",
              nodeId(), var()));
    }

    @Override
    public <NewT> SdkBindingData<NewT> as(
        SdkLiteralType<NewT> newType, Function<T, NewT> castFunction) {
      return create(newType, nodeId(), var());
    }

    @Override
    public final String toString() {
      return String.format("SdkBindingData{type=%s, nodeIs=%s, var=%s}", type(), nodeId(), var());
    }
  }

  @AutoValue
  abstract static class BindingCollection<T> extends SdkBindingData<List<T>> {
    abstract List<SdkBindingData<T>> bindingCollection();

    private static <T> BindingCollection<T> create(
        SdkLiteralType<T> elementType, List<SdkBindingData<T>> bindingCollection) {
      checkIncompatibleTypes(elementType, bindingCollection);
      return new AutoValue_SdkBindingData_BindingCollection<>(
          collections(elementType), bindingCollection);
    }

    @Override
    BindingData idl() {
      return BindingData.ofCollection(
          bindingCollection().stream().map(SdkBindingData::idl).collect(toUnmodifiableList()));
    }

    @Override
    public List<T> get() {
      return bindingCollection().stream().map(SdkBindingData::get).collect(toUnmodifiableList());
    }

    @Override
    public <NewT> SdkBindingData<NewT> as(
        SdkLiteralType<NewT> newElementType, Function<List<T>, NewT> castFunction) {
      throw new UnsupportedOperationException(
          "SdkBindingData of binding collections cannot be casted");
    }

    @Override
    public final String toString() {
      return String.format("SdkBindingData{type=%s, collection=%s}", type(), bindingCollection());
    }
  }

  @AutoValue
  public abstract static class BindingMap<T> extends SdkBindingData<Map<String, T>> {
    abstract Map<String, SdkBindingData<T>> bindingMap();

    private static <T> BindingMap<T> create(
        SdkLiteralType<T> valuesType, Map<String, SdkBindingData<T>> bindingMap) {
      checkIncompatibleTypes(valuesType, bindingMap.values());
      return new AutoValue_SdkBindingData_BindingMap<>(maps(valuesType), bindingMap);
    }

    @Override
    BindingData idl() {
      return BindingData.ofMap(
          bindingMap().entrySet().stream()
              .collect(toUnmodifiableMap(Map.Entry::getKey, e -> e.getValue().idl())));
    }

    @Override
    public Map<String, T> get() {
      return bindingMap().entrySet().stream()
          .collect(toUnmodifiableMap(Map.Entry::getKey, e -> e.getValue().get()));
    }

    @Override
    public <NewT> SdkBindingData<NewT> as(
        SdkLiteralType<NewT> newType, Function<Map<String, T>, NewT> castFunction) {
      throw new UnsupportedOperationException("SdkBindingData of binding map cannot be casted");
    }

    @Override
    public final String toString() {
      return String.format("SdkBindingData{type=%s, map=%s}", type(), bindingMap());
    }
  }

  private static <T> void checkIncompatibleTypes(
      SdkLiteralType<T> elementType, Collection<SdkBindingData<T>> elements) {
    List<LiteralType> incompatibleTypes =
        elements.stream()
            .map(SdkBindingData::type)
            .filter(type -> !type.equals(elementType))
            .map(SdkLiteralType::getLiteralType)
            .distinct()
            .collect(toList());
    if (!incompatibleTypes.isEmpty()) {
      throw new IllegalArgumentException(
          String.format(
              "Type mismatch: expected all elements of type %s but found some elements of type: %s",
              elementType.getLiteralType(), incompatibleTypes));
    }
  }
}

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingData.scala
/*
 * Copyright 2021 Flyte Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.flyte.flytekitscala

import org.flyte.api.v1.{
  BindingData,
  LiteralType,
  Primitive,
  Scalar,
  SimpleType
}
import org.flyte.flytekit.{SdkBindingData => SdkJavaBindinigData}

import java.time.{Duration, Instant}
import scala.collection.JavaConverters._

/** Utility to create [[SdkBindingData]] using scala raw types.
  */
object SdkBindingData {

  /** Creates a [[SdkBindingData]] for a flyte string ([[String]] for scala)
    * with the given value.
    *
    * @param string
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofString(string: String): SdkJavaBindinigData[String] =
    createSdkBindingData(string)

  /** Creates a [[SdkBindingData]] for a flyte integer ([[Long]] for scala) with
    * the given value.
    *
    * @param long
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofInteger(long: Long): SdkJavaBindinigData[Long] =
    createSdkBindingData(long)

  /** Creates a [[SdkBindingData]] for a flyte float ([[Double]] for scala) with
    * the given value.
    *
    * @param double
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofFloat(double: Double): SdkJavaBindinigData[Double] =
    createSdkBindingData(double)

  /** Creates a [[SdkBindingData]] for a flyte boolean ([[Boolean]] for scala)
    * with the given value.
    *
    * @param boolean
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofBoolean(
      boolean: Boolean
  ): SdkJavaBindinigData[Boolean] =
    createSdkBindingData(boolean)

  /** Creates a [[SdkBindingData]] for a flyte instant ([[Instant]] for scala)
    * with the given value.
    *
    * @param instant
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofDateTime(instant: Instant): SdkJavaBindinigData[Instant] =
    createSdkBindingData(instant)

  /** Creates a [[SdkBindingData]] for a flyte duration ([[Duration]] for scala)
    * with the given value.
    *
    * @param duration
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofDuration(
      duration: Duration
  ): SdkJavaBindinigData[Duration] = createSdkBindingData(duration)

  /** Creates a [[SdkBindingData]] for a flyte collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofCollection[T](
      collection: List[T]
  ): SdkJavaBindinigData[List[T]] = createSdkBindingData(collection)

  /** Creates a [[SdkBindingData]] for a flyte collection given a scala
    * [[List]].
    *
    * @param literalType
    *   literal type for the whole collection. It must be a
    *   [[LiteralType.Kind.COLLECTION_TYPE]].
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofCollection[T](
      literalType: LiteralType,
      collection: List[T]
  ): SdkJavaBindinigData[List[T]] =
    createSdkBindingData(collection, Option(literalType))

  /** Creates a [[SdkBindingData]] for a flyte string collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofStringCollection(
      collection: List[String]
  ): SdkJavaBindinigData[List[String]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.STRING)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte integer collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofIntegerCollection(
      collection: List[Long]
  ): SdkJavaBindinigData[List[Long]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.INTEGER)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte boolean collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofBooleanCollection(
      collection: List[Boolean]
  ): SdkJavaBindinigData[List[Boolean]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.BOOLEAN)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte float collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofFloatCollection(
      collection: List[Double]
  ): SdkJavaBindinigData[List[Double]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(LiteralType.ofSimpleType(SimpleType.FLOAT))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte datetime collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofInstantCollection(
      collection: List[Instant]
  ): SdkJavaBindinigData[List[Instant]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.DATETIME)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte duration collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofDurationCollection(
      collection: List[Duration]
  ): SdkJavaBindinigData[List[Duration]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.DURATION)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofMap[T](
      map: Map[String, T]
  ): SdkJavaBindinigData[Map[String, T]] = createSdkBindingData(map)

  /** Creates a [[SdkBindingData]] for a flyte string map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofStringMap(
      map: Map[String, String]
  ): SdkJavaBindinigData[Map[String, String]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.STRING))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte long map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofIntegerMap(
      map: Map[String, Long]
  ): SdkJavaBindinigData[Map[String, Long]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.INTEGER))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte boolean map given a scala
    * [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofBooleanMap(
      map: Map[String, Boolean]
  ): SdkJavaBindinigData[Map[String, Boolean]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.BOOLEAN))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte double map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofFloatMap(
      map: Map[String, Double]
  ): SdkJavaBindinigData[Map[String, Double]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.FLOAT))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte instant map given a scala
    * [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofInstantMap(
      map: Map[String, Instant]
  ): SdkJavaBindinigData[Map[String, Instant]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(
          LiteralType.ofSimpleType(SimpleType.DATETIME)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte duration map given a scala
    * [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofDurationMap(
      map: Map[String, Duration]
  ): SdkJavaBindinigData[Map[String, Duration]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(
          LiteralType.ofSimpleType(SimpleType.DURATION)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte duration map given a scala
    * [[Map]].
    *
    * @param literalType
    *   literal type for the whole collection. It must be a
    *   [[LiteralType.Kind.MAP_VALUE_TYPE]].
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofMap[T](
      literalType: LiteralType,
      map: Map[String, T]
  ): SdkJavaBindinigData[Map[String, T]] =
    createSdkBindingData(map, Option(literalType))

  private def toBindingData(
      value: Any,
      literalTypeOpt: Option[LiteralType]
  ): (BindingData, LiteralType) = {
    value match {
      case string: String =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofStringValue(string))
          ),
          LiteralType.ofSimpleType(SimpleType.STRING)
        )
      case boolean: Boolean =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofBooleanValue(boolean))
          ),
          LiteralType.ofSimpleType(SimpleType.BOOLEAN)
        )
      case long: Long =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofIntegerValue(long))
          ),
          LiteralType.ofSimpleType(SimpleType.INTEGER)
        )
      case double: Double =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofFloatValue(double))
          ),
          LiteralType.ofSimpleType(SimpleType.FLOAT)
        )
      case instant: Instant =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofDatetime(instant))
          ),
          LiteralType.ofSimpleType(SimpleType.DATETIME)
        )
      case duration: Duration =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofDuration(duration))
          ),
          LiteralType.ofSimpleType(SimpleType.DURATION)
        )
      case list: Seq[_] =>
        val literalType = literalTypeOpt.getOrElse {
          val (_, innerLiteralType) = toBindingData(
            list.headOption.getOrElse(
              throw new RuntimeException(
                "Can't create binding for an empty list without knowing the type, use SdkBindingData.of<type>Collection(...)"
              )
            ),
            literalTypeOpt = None
          )

          LiteralType.ofCollectionType(innerLiteralType)
        }

        (
          BindingData.ofCollection(
            list
              .map { innerValue =>
                val (bindingData, _) = toBindingData(innerValue, literalTypeOpt)
                bindingData
              }
              .toList
              .asJava
          ),
          literalType
        )
      case map: Map[String, _] =>
        val literalType = literalTypeOpt.getOrElse {
          val (_, innerLiteralType) = toBindingData(
            map.headOption
              .map(_._2)
              .getOrElse(
                throw new RuntimeException(
                  "Can't create binding for an empty map without knowing the type, use SdkBindingData.of<type>Map(...)"
                )
              ),
            literalTypeOpt = None
          )

          LiteralType.ofMapValueType(innerLiteralType)
        }
        (
          BindingData.ofMap(
            map
              .mapValues { innerValue =>
                val (bindingData, _) = toBindingData(innerValue, literalTypeOpt)
                bindingData
              }
              .toMap
              .asJava
          ),
          literalType
        )
      case other =>
        throw new IllegalStateException(
          s"${other.getClass.getSimpleName} class is not supported as SdkBindingData inner class"
        )
    }
  }

  private def createSdkBindingData[T](
      value: T,
      literalTypeOpt: Option[LiteralType] = None
  ): SdkJavaBindinigData[T] = {
    val (bindingData, literalType) = toBindingData(value, literalTypeOpt)
    SdkJavaBindinigData.create(bindingData, literalType, value)
  }
}
